File: app/src/main/java/com/faendir/lightning_launcher/multitool/proxy/Item.java
package com.faendir.lightning_launcher.multitool.proxy;

import androidx.annotation.StringDef;
import android.view.View;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;

/**
 * @author lukas
 * @since 04.07.18
 */
public interface Item extends Proxy {
    String TYPE_SHORTCUT = "Shortcut";
    String TYPE_FOLDER = "Folder";
    String TYPE_PANEL = "Panel";

    String getName();

    void setName(String name);

    int getId();

    String getTag();

    String getTag(String id);

    @Type
    String getType();

    int getWidth();

    int getHeight();

    float getPositionX();

    float getPositionY();

    float getScaleX();

    float getScaleY();

    float getRotation();

    void setRotation(float angle);

    RectL getCell();

    boolean isVisible();

    Image getBoxBackground(@Box.Mode String state);

    PropertySet getProperties();

    void setSize(float width, float height);

    Container getParent();

    void setTag(String id, String value);

    void setCell(int left, int top, int right, int bottom);

    void setCell(int left, int top, int right, int bottom, boolean portrait);

    void setPosition(float x, float y);

    void setScale(float scaleX, float scaleY);

    void setSkew(float skewX, float skewY);

    void setVisibility(boolean visible);

    void setBoxBackground(Image image, @Box.Mode String state, boolean persistent);

    void setBinding(@PropertySet.BindingProperty String target, String formula, boolean enabled);

    View getRootView();

    @Retention(RetentionPolicy.SOURCE)
    @StringDef({TYPE_SHORTCUT, TYPE_FOLDER, TYPE_PANEL})
    @interface Type {
    }
}

File: app/src/main/java/com/faendir/lightning_launcher/multitool/proxy/Item.kt
package com.faendir.lightning_launcher.multitool.proxy

import android.view.View
import androidx.annotation.StringDef

/**
 * @author lukas
 * @since 04.07.18
 */
interface Item : Proxy {

    var name: String

    val id: Int

    val tag: String

    @get:Type
    val type: String

    val width: Int

    val height: Int

    val positionX: Float

    val positionY: Float

    val scaleX: Float

    val scaleY: Float

    var rotation: Float

    val cell: RectL

    val isVisible: Boolean

    val properties: PropertySet

    val parent: Container

    val rootView: View

    fun getTag(id: String): String?

    fun getBoxBackground(@Box.Mode state: String): Image

    fun setSize(width: Float, height: Float)

    fun setTag(id: String, value: String?)

    fun setCell(left: Int, top: Int, right: Int, bottom: Int)

    fun setCell(left: Int, top: Int, right: Int, bottom: Int, portrait: Boolean)

    fun setPosition(x: Float, y: Float)

    fun setScale(scaleX: Float, scaleY: Float)

    fun setSkew(skewX: Float, skewY: Float)

    fun setVisibility(visible: Boolean)

    fun setBoxBackground(image: Image, @Box.Mode state: String, persistent: Boolean)

    fun setBinding(@PropertySet.BindingProperty target: String, formula: String, enabled: Boolean)

    @Retention(AnnotationRetention.SOURCE)
    @StringDef(TYPE_SHORTCUT, TYPE_FOLDER, TYPE_PANEL)
    annotation class Type

    companion object {
        const val TYPE_SHORTCUT = "Shortcut"
        const val TYPE_FOLDER = "Folder"
        const val TYPE_PANEL = "Panel"
    }
}
